src/passes/OptimizeInstructions.cpp

```diff
@@ -1677,6 +1719,34 @@ struct OptimizeInstructions
           }
         }
       }
+      // signed(x - y) <=> 0  =>  x <=> y
+      //
+      // unsigned(x - y) > 0    =>   x != y
+      // unsigned(x - y) <= 0   =>   x == y
```

Comment:
Would it be possible to do these with matchers as well?

Reply:
It could in future. Currently it little bit non-optimal without something like 
```cpp
matches(&curr, binary(
  or(Abstact::Gt, Abstact::Le, ...)
  , ...))
```

Reply:
Oh right, sorry I haven't been able to get to implementing that yet. Do the matches for those patterns fit on one line? If so, I think the repetition would still be a readability improvement. Otherwise we can keep this as-is for now.

Reply:
Hmm, I'm not sure this is much better:
```cpp
{
        using namespace Match;

        BinaryOp op;
        Binary* inner;

        // unsigned(x - y) > 0    =>   x != y
        if (matches(curr,
                    binary(Abstract::GtU,
                           binary(&inner, Abstract::Sub, any(), any()), ival(0)))) {
          curr->op = Abstract::getBinary(type, Abstract::Ne);
          curr->right = inner->right;
          curr->left = inner->left;
          return curr;
        }

        // unsigned(x - y) <= 0   =>   x == y
        if (matches(curr,
                    binary(Abstract::LeU,
                           binary(&inner, Abstract::Sub, any(), any()), ival(0)))) {
          curr->op = Abstract::getBinary(type, Abstract::Eq);
          curr->right = inner->right;
          curr->left = inner->left;
          return curr;
        }

        // for signed or sign-agnostic relationals we don't change op.
        // x - y <=> 0  =>  x <=> y
        if (matches(curr,
                    binary(&op,
                           binary(&inner, Abstract::Sub, any(), any()), ival(0))) &&
            (op == Abstract::getBinary(type, Abstract::Eq) ||
             op == Abstract::getBinary(type, Abstract::Ne) ||
             op == Abstract::getBinary(type, Abstract::LeS) ||
             op == Abstract::getBinary(type, Abstract::LtS) ||
             op == Abstract::getBinary(type, Abstract::GeS) ||
             op == Abstract::getBinary(type, Abstract::GtS))) {
          curr->right = inner->right;
          curr->left = inner->left;
          return curr;
        }
      }
```
I see this much more verbose and less efficient

Reply:
This actually looks much more readable to me (and will be even better after #3284).

Reply:
Alright. I refactored this